File: 033. Search in Rotated Sorted Array/binarysearch.cpp
class Solution {
public:
    int search(vector<int>& v, int target) {
        if(v.size() == 0) 
            return -1;
        int len = v.size(), low = 0, high = len - 1;
        while(low <= high){
            int mid = (low + high) / 2;
            // cout << "low:" << low << ", high:" << high << ", mid:" << mid << endl;
            int cur = v[mid];
            if(cur == target){
                return mid;
            }else if(cur < target){
                // mid in the original part
                if(cur >= v[low]){
                    low = mid + 1;
                }else{
                    // mid in the rotated part
                    if(target > v[high]){
                        // target in the original part
                        high = mid - 1;
                    }else{
                        // target in the rotated part
                        low = mid + 1;
                    }
                }
            }else if(cur > target){
                // mid in the rotated part
                if(cur <= v[high]){
                    high = mid - 1;
                }else{
                    // mid in the original part
                    if(target >= v[low]){
                        // target in the original part
                        high = mid - 1;
                    }else{
                        // target in the rotated part
                        low = mid + 1;
                    }
                }
            }
        }

         return -1;
    }
};

File: 0081. Search in Rotated Sorted Array II/binarysearch.java
class Solution {
    public boolean search(int[] v, int target) {
        int len = v.length, low = 0, high = len - 1;
        while(low <= high){
            System.out.println(low + "," + high);
            while(low < high && v[low] == v[low + 1]){
                low++;
            }
            while(low < high && v[high] == v[high - 1]){
                high--;
            }
            int mid = (low + high) / 2, cur = v[mid];
            if(cur == target)
                return true;
            else if(cur < target){
                if(cur >= v[low]){
                    low = mid + 1;
                }else{
                    if(v[high] >= target){
                        low = mid + 1;
                    }else{
                        high = mid - 1;
                    }
                }
            }else{// cur > target
                if(cur <= v[high]){
                    high = mid - 1;
                }else{
                    if(target >= v[low]){
                        high = mid - 1;
                    }else{
                        low = mid + 1;
                    }
                }
            }
        }
        return false;
        
    }
}